### PR TITLE
python server

### PR DESCRIPTION
**Pull request by opreateodora**

### CODE REVIEW

File: runtime-management/runtime-management-impl/pom.xml [hunk 1]
```diff
@@ -122,6 +122,12 @@
             <artifactId>caffeine</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>org.jboss.resteasy</groupId>
+            <artifactId>resteasy-client</artifactId>
+            <version>4.7.2.Final</version>
```
Comment (razvanfulea) [inline runtime-management/runtime-management-impl/pom.xml:128]: Extract version into a higher level pom

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        return getPythonEvaluationResult(script, prepareEnvironmentScript, vars);
+    }
+
+    private PythonEvaluationResult getPythonEvaluationResult(String expression, String prepareEnvironmentScript,
+                                                             Map<String, Serializable> context) throws JsonProcessingException {
+        String payload = generatePayloadForEval(expression, prepareEnvironmentScript, context);
+        EvaluationResults scriptResults = executeRequestOnPythonServer("POST", payload);
+
+        String exception = scriptResults.getException();
+        if (!StringUtils.isEmpty(exception)) {
```
Comment (razvanfulea) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java:76]: if (StringUtils.isNotEmpty(exception)) {

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            logger.error(String.format("Failed to execute script {%s}", exception));
+            throw new ExternalPythonEvalException(exception);
+        }
+        context.put("accessed_resources_set", (Serializable) scriptResults.getAccessedResources());
```
Comment (razvanfulea) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java:80]: Consider extracting accessed_resources_set into a constant

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        String encodedAuth = getEncoder().encodeToString((username + ":" + password).getBytes(UTF_8));
+        return "Basic " + encodedAuth;
+    }
+
+    private String generatePayloadForEval(String expression, String prepareEnvironmentScript,
+                                          Map<String, Serializable> context) throws JsonProcessingException {
+        HashMap<String, Serializable> payload = new HashMap<>(4);
+        payload.put("expression", expression);
+        payload.put("envSetup", prepareEnvironmentScript);
+        payload.put("context", (Serializable) context);
```
Comment (razvanfulea) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java:111]: Why new HashMap<>(3) if we only add 3 props?

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                .invoke();
+
+        return objectMapper.readValue(scriptResponse.readEntity(String.class), EvaluationResults.class);
+    }
+
+    private String getBasicAuthorizationHeaderValue() {
+        String username = "administrator";
+        String password = "password";
+        String encodedAuth = getEncoder().encodeToString((username + ":" + password).getBytes(UTF_8));
+        return "Basic " + encodedAuth;
```
Comment (razvanfulea) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java:103]: Consider extracting these as system properties

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonRuntimeServiceImpl.java [hunk 6]
```diff
@@ -76,13 +83,40 @@ public PythonExecutionResult exec(Set<String> dependencies, String script, Map<S
...
-            throw new ExternalPythonScriptException("Execution was interrupted while waiting for a python permit.");
+        if (PYTHON_SERVER.equals(PYTHON_EVALUATOR)) {
+            try {
+                return externalPythonServerService.evalOnExternalPythonServer(script, prepareEnvironmentScript, vars);
+            } catch (JsonProcessingException ie) {
+                logger.error(ie);
+                throw new ExternalPythonScriptException("Execution was interrupted while waiting for a python permit.");
+            }
+
+            //case PYTHON.equals(PYTHON_EVALUATOR)
```
Comment (razvanfulea) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonRuntimeServiceImpl.java:94]: remove comment

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonRuntimeServiceImpl.java [hunk 7]
```diff
@@ -76,13 +83,40 @@ public PythonExecutionResult exec(Set<String> dependencies, String script, Map<S
...
+                    try {
+                        logger.info("Acquired a permit for a new python process. Continuing with execution...");
+                        return externalPythonExecutionEngine.eval(prepareEnvironmentScript, script, vars);
+                    } finally {
+                        executionControlSemaphore.release();
+                    }
+                }
+            } catch (InterruptedException ie) {
+                throw new ExternalPythonScriptException("Execution was interrupted while waiting for a python permit.");
+            }
```
Comment (razvanfulea) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonRuntimeServiceImpl.java:117]: extract this in a separate function

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java [hunk 8]
```diff
@@ -47,7 +47,7 @@
 public class ExternalPythonServerServiceImpl implements ExternalPythonServerService {
 
     private static final Logger logger = LogManager.getLogger(ExternalPythonServerServiceImpl.class);
-    private static final String EXTERNAL_PYTHON_PORT = System.getProperty("python.port");
+    private static final String EXTERNAL_PYTHON_PORT = getPythonPort(System.getProperty("python.port"));
```
Comment (razvanfulea) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java:53]: `System.getProperty("python.port", 8001)` does the same thing

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java [hunk 9]
```diff
@@ -105,13 +106,20 @@ private String getBasicAuthorizationHeaderValue() {
...
         return objectMapper.writeValueAsString(payload);
     }
 
+    private static String getPythonPort(String pythonPort) {
+        String defaultPort = "8001";
+        if (pythonPort != null)
+            return pythonPort;
+        return defaultPort;
+    }
+
```
Comment (razvanfulea) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java:138]: remove this function

File: runtime-management/runtime-management-impl/src/main/resources/pythonServer.properties [hunk 10]
```diff
@@ -0,0 +1,2 @@
+username=administrator
+password=password
```
Comment (razvanfulea) [inline runtime-management/runtime-management-impl/src/main/resources/pythonServer.properties:2]: You can leave these 2 values blank

File: runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+import static org.jboss.resteasy.util.HttpHeaderNames.CONTENT_TYPE;
+import static org.springframework.util.MimeTypeUtils.APPLICATION_JSON;
+
+@Service
+public class ExternalPythonServerServiceImpl implements ExternalPythonServerService {
+
+    private static final Logger logger = LogManager.getLogger(ExternalPythonServerServiceImpl.class);
+    private static final String EXTERNAL_PYTHON_PORT = System.getProperty("python.port", String.valueOf(8001));
+    private static final String EXTERNAL_PYTHON_SERVER_URL = "https://localhost:" + EXTERNAL_PYTHON_PORT;
+    private static final String EXTERNAL_PYTHON_SERVER_EVAL_PATH = "/worker/rest/v1/expressions";
```
Comment (pentektamas) [inline runtime-management/runtime-management-impl/src/main/java/io/cloudslang/runtime/impl/python/external/ExternalPythonServerServiceImpl.java:55]: Change to the new path: /rest/v1/eval

**Pull request merged**